README.md
# nu-react-native-acss

[![npm package][npm-badge]][npm]

[npm-badge]: https://img.shields.io/npm/v/npm-package.png?style=flat-square
[npm]: https://www.npmjs.org/package/@_nu/react-native-acss

Quick way to write style in react native.

Idea from Atomic CSS Design system.

```
$ yarn add @_nu/react-native-acss
```

## How to use?

```JSX
import React from 'react';
import {Text, View, ImageBackground, Image} from './packages/withAcss';
import mImg1 from "./assets/1.jpg";
import mImg2 from "./assets/2.jpg";

export default function App() {
    return (
        <View bc="#f5f5f5" pt={100} ph={30} h="100%">
            <View mb={16} aic>
                <Image br={32} w={64} h={64} source={mImg1}/>
            </View>
            <Text c="#ffffff" p={16} bc="#4c5fe2" tac mb={16}>Hello world!</Text>
            <ImageBackground source={mImg2} mb={40}>
                <View bc="rgba(255,255,255,.8)">
                    <Text fs={40} lh={100} tac fwb c="#4c5fe2">Hello Again!</Text>
                </View>
            </ImageBackground>
        </View>
    );
};
```

## How to bind to custom component?

```JSX
import React from 'react';
import {Text, View} from 'react-native';
import widthAcss from './packages/withAcss';

const CusText = widthAcss(Text);
const CusView = widthAcss(View);

export default function App() {
    return (
        <CusView bc="#4c5fe2" p={16}>
            <CusText tac c="#ffffff">Custom text</CusText>
        </CusView>
    );
};
```

![demo](./demo.png)

## Name rules

It looks wired in first sight what we use abbreviation instead of the full name. Cause we want user just remember the rules below rather than all the props.

1. Only Initials: `aic` means `alignItems: center`;
2. Direct digital connection: `fw700` means `fontWidth:700`;
4. `p` means `percent`: `w100p` means `width:'100%'`;

```JSX
<View mb={16} bc="#4c5fe2" aic>
    <Text c="#ffffff" pt={16} pb={16} tac>Hello world!</Text>
</View>
```

equal to

```JSX
<View style={{
    marginBottom:16,
    backgroundColor:'#4c5fe2',
    alignItem:'center'
}}>
    <Text style={{
      color:'#ffffff',
      paddingTop: 16,
      paddingBottom: 16,
      alignItem:'center'
    }}>Hello world!</Text>
</View>
```

## API

```JSX
const styleMap = {
  ac: 'alignContent',
  acs: ['alignContent', 'stretch'],
  acc: ['alignContent', 'center'],
  acfs: ['alignContent', 'flex-start'],
  acfe: ['alignContent', 'flex-end'],
  acsb: ['alignContent', 'space-between'],
  acsa: ['alignContent', 'space-around'],
  ai: 'align-items',
  ais: ['alignItems', 'stretch'],
  aib: ['alignItems', 'baseline'],
  aic: ['alignItems', 'center'],
  aifs: ['alignItems', 'flex-start'],
  aife: ['alignItems', 'flex-end'],
  as: 'alignSelf',
  asa: ['alignSelf', 'auto'],
  ass: ['alignSelf', 'stretch'],
  asb: ['alignSelf', 'baseline'],
  asc: ['alignSelf', 'center'],
  asfs: ['alignSelf', 'flex-start'],
  asfe: ['alignSelf', 'flex-end'],
  bc: 'backgroundColor',
  br: 'borderRadius',
  bw: 'borderWidth',
  bs: 'borderStyle',
  btw: 'borderTopWidth',
  btc: 'borderTopColor',
  brw: 'borderRightWidth',
  brc: 'borderRightColor',
  bbw: 'borderBottomWidth',
  bbc: 'borderBottomColor',
  blw: 'borderLeftWidth',
  blc: 'borderLeftColor',
  btrr: 'borderTopRightRadius',
  btlr: 'borderTopLeftRadius',
  bbrr: 'borderBottomRightRadius',
  bblr: 'borderBottomLeftRadius',
  b: 'bottom',
  b0: 'bottom',
  c: 'color',
  dn: ['display', 'none'],
  f: 'flex',
  f1: ['flex', 1],
  fdr: ['flexDirection', 'row'],
  fdrr: ['flexDirection', 'row-reverse'],
  fdc: ['flexDirection', 'column'],
  fdcr: ['flexDirection', 'column-reverse'],
  fwn: ['flexWrap', 'nowrap'],
  fwb: ['fontWeight', 'bold'],
  fw100: ['fontWeight', 100],
  fw200: ['fontWeight', 200],
  fw300: ['fontWeight', 300],
  fw400: ['fontWeight', 400],
  fw500: ['fontWeight', 500],
  fw600: ['fontWeight', 600],
  fw700: ['fontWeight', 700],
  fw800: ['fontWeight', 800],
  fw900: ['fontWeight', 900],
  fs: 'fontSize',
  ff: 'fontFamily',
  fsn: ['fontStyle', 'normal'],
  fsi: ['fontStyle', 'italic'],
  h: 'height',
  h100p: ['height', '100%'],
  jcc: ['justifyContent', 'center'],
  jcsb: ['justifyContent', 'space-between'],
  jcfs: ['justifyContent', 'flex-start'],
  jcfe: ['justifyContent', 'flex-end'],
  jcsa: ['justifyContent', 'space-around'],
  jcse: ['justifyContent', 'space-evenly'],
  l: 'left',
  l0: ['left', 0],
  lh: 'lineHeight',
  m(value) {
    return {
      marginTop: value,
      marginBottom: value,
      marginRight: value,
      marginLeft: value,
    };
  },
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  mh: 'marginHorizontal',
  mv: 'marginVertical',
  o: 'opacity',
  oh: ['overflow', 'hidden'],
  p(value) {
    return {
      paddingTop: value,
      paddingBottom: value,
      paddingRight: value,
      paddingLeft: value,
    };
  },
  pa: ['position', 'absolute'],
  pr(value) {
    if (value === true) {
      return {
        position: 'relative',
      };
    }
    return {
      paddingRight: value,
    };
  },
  pt: 'paddingTop',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
  ph: 'paddingHorizontal',
  pv: 'paddingVertical',
  r: 'right',
  r0: ['right', 0],
  t: 'top',
  t0: ['top', 0],
  tac: ['textAlign', 'center'],
  tar: ['textAlign', 'right'],
  taa: ['textAlign', 'auto'],
  tal: ['textAlign', 'left'],
  taj: ['textAlign', 'justify'],
  ttu: ['textTransform', 'uppercase'],
  ttc: ['textTransform', 'capitalize'],
  ttn: ['textTransform', 'none'],
  ttl: ['textTransform', 'lowercase'],
  w: 'width',
  w100p: ['width', '100%'],
  zi: 'zIndex',
};
```

Pure, simple, natural. 

As you can see that is all the code we create blew. 

We just did a little but can help you a lot when create style in react native.

help us to make it better together [issue](https://github.com/nu-system/acss-react-native/issues).


CHANGELOG.md
## 0.0.3 (2019-07-22)


### Features

* 移除 style-component ([8f01618](https://github.com/nu-system/react-native-acss/commit/8f01618))





__tests__/index.test.js
import React from 'react';
import  {
  Text, View, ImageBackground, Image,
} from '../lib/index';
import {shallow,mount} from 'enzyme';

import img from "./assets/1.jpg"

describe("Text",()=>{
    it(`Text正常渲染`,()=>{
        const tmp= shallow(<Text>Hello Again!</Text>)
        expect(tmp.find('Text')).toHaveLength(1);
    })
    it(`Text Acss 样式渲染`,()=>{
        const expectRes={
            fontSize: 40,
            lineHeight: 100,
            textAlign: 'center',
            fontWeight: 'bold',
            color: '#4c5fe2' 
        }
        const tmp= shallow(<Text fs={40} lh={100} tac fwb c="#4c5fe2">Hello Again!</Text>)
        const style=tmp.prop("style")
        for(let i in style){
            expect(style[i]).toBe(expectRes[i])
        }
    })
})

describe("View",()=>{
    it(`View正常渲染`,()=>{
        const tmp= shallow(<View></View>)
        expect(tmp.find('View')).toHaveLength(1);
    })
    it(`View Acss 样式渲染`,()=>{
        const expectRes={
            marginBottom: 16,
            alignItems: 'center'
        }
        const tmp= shallow(<View mb={16} aic></View>)
        const style=tmp.prop("style")
        for(let i in style){
            expect(style[i]).toBe(expectRes[i])
        }
    })
})

describe("ImageBackground",()=>{
    it(`ImageBackground正常渲染`,()=>{
        const tmp= shallow(<ImageBackground></ImageBackground>)
        expect(tmp.find('ImageBackground')).toHaveLength(1);
    })
    it(`ImageBackground Acss 样式渲染`,()=>{
        const expectRes={
            marginBottom: 40
        }
        const tmp= shallow(<ImageBackground source={img} mb={40}></ImageBackground>)
        const style=tmp.prop("style")
        for(let i in style){
            expect(style[i]).toBe(expectRes[i])
        }
    })
})

describe("Image",()=>{
    it(`Image正常渲染`,()=>{
        const tmp= shallow(<Image></Image>)
        expect(tmp.find('Image')).toHaveLength(1);
    })
    it(`Image Acss 样式渲染`,()=>{
        const expectRes={
            borderRadius: 32,
            width:64,
            height:64
        }
        const tmp= shallow(<Image br={32} w={64} h={64} source={img} />)
        const style=tmp.prop("style")
        for(let i in style){
            expect(style[i]).toBe(expectRes[i])
        }
    })
})


__tests__/sattr.test.js
import React from 'react';
import  {
  Text, View, ImageBackground, Image,
} from '../lib/index';
import {shallow,mount} from 'enzyme';

/** 
 * 此处测试方法的属性
*/
describe("function attribute",()=>{
    it("p padding四个方向",()=>{
        const expectRes={
            paddingTop:40,
            paddingLeft:40,
            paddingBottom:40,
            paddingRight:40
        }
        const tmp=shallow(<View p={40}></View>)
        const style=tmp.prop("style")
        for(let i in style){
            expect(style[i]).toBe(expectRes[i])
        }
    })
    it("m margin4个方法向",()=>{
        const expectRes={
            marginTop:40,
            marginLeft:40,
            marginBottom:40,
            marginRight:40
        }
        const tmp=shallow(<View m={40}></View>)
        const style=tmp.prop("style")
        for(let i in style){
            expect(style[i]).toBe(expectRes[i])
        }
    })
    it("pr 两种可能 —— position:relative",()=>{
        const expectRes={
            position:"relative"
        }
        const tmp=shallow(<View pr></View>)
        const style=tmp.prop("style")
        for(let i in style){
            expect(style[i]).toBe(expectRes[i])
        }
    })
    it("pr 两种可能 —— padding-right",()=>{
        const expectRes={
            paddingRight:40
        }
        const tmp=shallow(<View pr={40}></View>)
        const style=tmp.prop("style")
        for(let i in style){
            expect(style[i]).toBe(expectRes[i])
        }
    })
})

__tests__/attribute.test.js

import {acssPropTypes
} from '../lib/index';
import styleMap from "../lib/styleMap";

function getAttr(str){
    if(typeof str==="string"){
        if(/%/.test(str)){
            return str.replace("%","p")
        }else if(parseInt(str)>0){
            return str;
        }else{
            let arr=str.match(/[A-Z]/g);
            if(arr){
                return str[0]+arr.join("").toLowerCase()
            }else{
                return str[0]
            }
        }
    }else{
        return str;
    }
}
describe("Attribute",()=>{
    const arrays={}
    const values={}
    for(let i in styleMap){
        if(typeof styleMap[i] === "string"){
            values[i]= getAttr(styleMap[i])
        }else if(styleMap[i] instanceof Array){
            arrays[i]=styleMap[i].map((s)=>{
                return getAttr(s)
            }).join("")
        }
    }
    for(let i in values){
        it(`${i}`,()=>{
            expect(i).toBe(values[i])
        })
    }
})
describe("map stylemap to proptypes",()=>{
    it("is proptypes sync with stylemap",()=>{
        
        for(let i in styleMap){
            expect(acssPropTypes[i] instanceof Function).toBe(true)
        }

    })
})


lib/index.js
import React from 'react';
import PropTypes from 'prop-types';
import {
  StyleSheet, View, Text, Image, ImageBackground,
} from 'react-native';
import styleMap from "./styleMap"

/**
 *
 * @type {{pr: ((function(): null)|shim), pt: ((function(): null)|shim), pv: ((function(): null)|shim), b0: shim, brc: ((function(): null)|shim), fw800: shim, fw400: shim, ac: ((function(): null)|shim), bbc: ((function(): null)|shim), brw: ((function(): null)|shim), ai: ((function(): null)|shim), zi: ((function(): null)|shim), as: ((function(): null)|shim), bbw: ((function(): null)|shim), aic: shim, aib: shim, btrr: ((function(): null)|shim), t0: shim, fsi: shim, fsn: shim, bc: ((function(): null)|shim), acfs: shim, l0: shim, ais: shim, jcc: shim, acfe: shim, br: ((function(): null)|shim), btc: ((function(): null)|shim), fw500: shim, bs: ((function(): null)|shim), fdrr: shim, asfs: shim, fw100: shim, bw: ((function(): null)|shim), blc: ((function(): null)|shim), asfe: shim, fw900: shim, b: ((function(): null)|shim), c: ((function(): null)|shim), btw: ((function(): null)|shim), fdc: shim, f: ((function(): null)|shim), h: ((function(): null)|shim), blw: ((function(): null)|shim), l: ((function(): null)|shim), m: ((function(): null)|shim), asa: shim, o: ((function(): null)|shim), asc: shim, p: ((function(): null)|shim), asb: shim, r: ((function(): null)|shim), t: ((function(): null)|shim), fdr: shim, btlr: ((function(): null)|shim), jcsa: shim, w: ((function(): null)|shim), jcsb: shim, jcse: shim, lh: ((function(): null)|shim), acc: shim, ass: shim, fdcr: shim, jcfe: shim, dn: shim, f1: shim, acs: shim, fw600: shim, fw200: shim, mb: ((function(): null)|shim), taa: shim, tac: shim, mh: ((function(): null)|shim), taj: shim, tal: shim, ml: ((function(): null)|shim), tar: shim, mr: ((function(): null)|shim), mt: ((function(): null)|shim), mv: ((function(): null)|shim), bbrr: ((function(): null)|shim), fwb: shim, fwn: shim, ff: shim, aife: shim, fs: ((function(): null)|shim), fw700: shim, fw300: shim, aifs: shim, oh: shim, acsb: shim, acsa: shim, ttc: shim, bblr: ((function(): null)|shim), ttl: shim, ttn: shim, pa: shim, pb: ((function(): null)|shim), ttu: shim, jcfs: shim, ph: ((function(): null)|shim), pl: ((function(): null)|shim), r0: shim}}
 */
/* eslint-disable */
const acssPropTypes = {};
for(let i in styleMap){
  if(styleMap[i] instanceof Array){
    acssPropTypes[i]=PropTypes.bool
  }else{
    acssPropTypes[i]=PropTypes.oneOfType([PropTypes.number, PropTypes.string])
  }
}
/* eslint-enable */

const getPropsWidthStyle = (props) => {
  const { style, ...otherProps } = props;
  const acssStyle = {};
  Object.keys(otherProps).forEach((key) => {
    const value = otherProps[key];
    const styleInfo = styleMap[key];
    if (value && styleInfo) {
      delete otherProps[key];
      if (typeof styleInfo === 'string') {
        acssStyle[styleInfo] = value;
      } else if (styleInfo instanceof Array && styleInfo.length === 2) {
        const [styleName, StyleValue] = styleInfo;
        acssStyle[styleName] = StyleValue;
      } else if (typeof styleInfo === 'function') {
        const ret = styleInfo(value);
        if (ret) {
          Object.assign(acssStyle, ret);
        }
      }
    }
  });
  otherProps.style = StyleSheet.flatten([acssStyle, style]);
  return otherProps;
};

const widthAcss = (Component) => {
  const Acss = props => <Component {...getPropsWidthStyle(props)} />;
  Acss.propTypes = acssPropTypes;
  return Acss;
};

const V = widthAcss(View);
const T = widthAcss(Text);
const ImgBg = widthAcss(ImageBackground);
const Img = widthAcss(Image);

V.propTypes = acssPropTypes;
T.propTypes = acssPropTypes;
ImgBg.propTypes = acssPropTypes;
Img.propTypes = acssPropTypes;

export default widthAcss;
export {
  V as View, T as Text, Img as Image, ImgBg as ImageBackground, acssPropTypes,
};


jest.config.js
// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after `n` failures
  // bail: 0,

  // Respect "browser" field in package.json when resolving modules
  // browser: false,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/private/var/folders/f3/p_8cqzpx6q35q4wkwbzdnf1r0000gn/T/jest_dx",

  // Automatically clear mock calls and instances between every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: null,

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // An array of regexp pattern strings used to skip coverage collection
  coveragePathIgnorePatterns: [
    "/node_modules/",
    "/demo/"
  ],

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: null,

  // A path to a custom dependency extractor
  // dependencyExtractor: null,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files using an array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: null,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: null,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   "js",
  //   "json",
  //   "jsx",
  //   "ts",
  //   "tsx",
  //   "node"
  // ],

  // A map from regular expressions to module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  modulePathIgnorePatterns: [
    "/demo/"
  ],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "failure-change",

  // A preset that is used as a base for Jest's configuration
  preset: 'react-native',

  // Run tests from one or more projects
  // projects: null,

  // Use this configuration option to add custom reporters to Jest
  reporters: ["default", "<rootDir>__tests__/my-custom-reporter.js"],

  // Automatically reset mock state between every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: null,

  // Automatically restore mock state between every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: null,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  setupFilesAfterEnv: ["<rootDir>__tests__/setup.js"],

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  testEnvironment: "jsdom",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  testMatch: [
    "**/__tests__/**/*.test.[jt]s?(x)",
    "**/?(*.)+(spec|test).[tj]s?(x)"
  ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  testPathIgnorePatterns: [
    "/node_modules/",
    "/demo/"
  ],

  // The regexp pattern or array of patterns that Jest uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: null,

  // This option allows use of a custom test runner
  // testRunner: "jasmine2",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "http://localhost",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  // transform: null,

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  transformIgnorePatterns: [
    "node_modules/(?!(jest-)?react-native)"
  ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: null,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};


demo/App.js
import React from 'react';
import { Text as RnText, View as RnView } from 'react-native';
import widthAcss, {
  Text, View, ImageBackground, Image,
} from './packages/withAcss';
import mImg1 from './assets/1.jpg';
import mImg2 from './assets/2.jpg';


const CusText = widthAcss(RnText);
const CusView = widthAcss(RnView);

export default function App() {
  return (
    <View bc="#f5f5f5" pt={100} ph={30} h="100%">
      <View mb={16} aic>
        <Image br={32} w={64} h={64} source={mImg1} />
      </View>
      <Text c="#ffffff" p={16} bc="#4c5fe2" tac mb={16}>Hello world!</Text>
      <ImageBackground source={mImg2} mb={40}>
        <View bc="rgba(255,255,255,.8)">
          <Text fs={40} lh={100} tac fwb c="#4c5fe2">Hello Again!</Text>
        </View>
      </ImageBackground>
      <CusView bc="#4c5fe2" p={16}>
        <CusText tac c="#ffffff">Custom text</CusText>
      </CusView>
    </View>
  );
}
